### ISSUE TITLE
HTML5 exports have hard coded use of #canvas (id = canvas) selector, setCanvas(element) from Engine does nothing

### ISSUE THREAD

**Issue by tuckersn** [godotengine/godot#44675]
Related to #44368 but not for the purpose of multiple instances of the engine running.

It's generally bad practice to use ids in modern web development, so ideally it wouldn't be necessary to use them in your Godot shell. This is especially problematic when trying to integrate into a framework like Angular or React where this breaks all conventions.

**Godot version:**
v3.2.3.stable.mono.official

**OS/device including version:**
Testing with Google Chrome

**Issue description:**
If you use any other id value besides "canvas" your HTML5 shell will not work. If you try to use no id value, your shell will not work. This is the error I got following the [documentation for setCanvas you can find here](https://docs.godotengine.org/en/stable/tutorials/platform/customizing_html5_shell.html#customizing-the-presentation)
![image](https://user-images.githubusercontent.com/8574134/103130038-de3f8e80-464f-11eb-9dff-eea7ca3cee26.png)

My understanding of the issue comes from the generated EMScripten JS file generated by the HTML5 export, after using a beautifier on the JS file, tracking down where the errors were occurring one by one, I eventually found the multiple references to `#canvas` when looking for why the targets of input events were consistently null whenever the canvas had an id other than `canvas` for input events on the canvas. 

The specific error above is triggered when registering an event handler for the mouse button, and if you comment out the line that break it breaks again for the mouse wheel a few hundred lines below that point. There are multiple hardcoded values, but they are global values such as `#window` and `#document` it's only the input handling for things going through the canvas element that are problematic. 

Example of one of the methods generated that use a hardcoded canvas id:
```js
function __findEventTarget(target) {
	try {
		if (!target) return window;
		if (typeof target === "number") target = __specialEventTargets[target] || UTF8ToString(target);
		if (target === "#window") return window;
		else if (target === "#document") return document;
		else if (target === "#screen") return screen;
		else if (target === "#canvas") return Module["canvas"];
		return typeof target === "string" ? document.getElementById(target) : target
	} catch (e) {
		return null
	}
}
```
Now if you use another id let's say called `#another-id` it won't work until you open the .JS file and find and repalce `"#canvas"` with `"another-id"`.

**Steps to reproduce:**
1. Take the default shell found at [https://github.com/godotengine/godot/blob/master/misc/dist/html/full-size.html](https://github.com/godotengine/godot/blob/master/misc/dist/html/full-size.html)
2. Change the <canvas id="canvas"> stuff to <canvas id="new-name">
3. Under `var engine = new Engine;` put `engine.setCanvas(document.getElementById("new-name"))`
4. It will return the error above.

**Minimal reproduction project:**
Various shells I created to test are in the shells folder, you can choose which to test by selecting runnable in the exports section, and then clicking the HTML5 icon in the top right. 

Shells tested in the reproduction.
- Default shell used
- [Documentation for setCanvas that doesn't work.](https://docs.godotengine.org/en/stable/tutorials/platform/customizing_html5_shell.html#customizing-the-presentation)
- Class Name Selection for setCavas using the same Id (Works because the id is still "canvas").
- Different Id Selection for setCanvas (Doesn't work because of the hardcoded use of  the "canvas" id).
- Working Shell is the controlled test, no setCanvas call.

Also attached is the default shell's 16,509 JS file, look up "#canvas" in it find every instance of this specific id being hardcoded.
[WebDemo.zip](https://github.com/godotengine/godot/files/5741693/WebDemo.zip)
[default-shell.zip](https://github.com/godotengine/godot/files/5741696/default-shell.zip)

**Comment by Calinou** [comment]
What should we use instead of IDs? I'm not sure if classes are ideal. Maybe data attributes?

**Comment by tuckersn** [comment]
> > It's generally bad practice to use ids in modern web development, so ideally it wouldn't be necessary to use them in your Godot shell. This is especially problematic when trying to integrate into a framework like Angular or React where this breaks all conventions.
> 
> What should we use instead of IDs? I'm not sure if classes are ideal since these are better used from CSS. Maybe data attributes?
> 
> Also, should we prefix all our class/attribute names with `godot-` to avoid conflicts with other libraries?

The way that setCanvas() is written in the [documentation](https://docs.godotengine.org/en/stable/tutorials/platform/html5_shell_classref.html#engine.setCanvas), we could just pass the actual element of the canvas to the Engine. So instead of looking for an id, class, or attribute, you just test if the targets are whatever was passed into setCanvas. This would make it a lot easier to integrate with things like jQuery or React and Angular which have style encapsulation, or other ways of selecting elements that aren't in frame-workless web development. 

Basically the person writing the "HTML Shell" would decide how the element is found by the Engine by just passing in the actual element to setCanvas instead of looking it up on it's own. I think classes or attributes would be better, but I think the way setCanvas is already written would be ideal.

**Comment by tuckersn** [comment]
Just tested on v3.2.4.beta4.mono.official

The behavior in the beta is different. You can use any id, but must have an id. You can't select by class / attribute / other without having *some* id on the element, but this is a lot better!
```ts
let canvas = document.getElementsByClassName("godot-canvas").item(0) as HTMLCanvasElement;
```

Works:
```html
<canvas class="canvas" id="test" width="400" height="400">
	HTML5 canvas appears to be unsupported in the current browser.<br />
	Please try updating or use a different browser.
</canvas>
```

Doesn't work:
```html
<canvas class="canvas" width="400" height="400">
	HTML5 canvas appears to be unsupported in the current browser.<br />
	Please try updating or use a different browser.
</canvas>
```

**Comment by Faless** [comment]
As you noticed, the issue was fixed in 3.2.4beta.

Yes, emscripten functions require canvases to have an ID, I don't see how this is a problem, you can also generate it on the fly if you prefer.

**Comment by tuckersn** [comment]
> As you noticed, the issue was fixed in 3.2.4beta.
> 
> Yes, emscripten functions require canvases to have an ID, I don't see how this is a problem, you can also generate it on the fly if you prefer.

Makes sense to me, shouldn't be an issue to randomly generate part of an id.

Thanks Faless! Sorry I hadn't checked the beta before posting.
